client/my-sites/stats/stats-post-anniversaries/index.jsx

```diff
@@ -0,0 +1,241 @@
+				},
+			} );
+		}
+
+		if ( period.period === 'month' ) {
+			return date.format( 'MMMM YYYY' );
+		}
+
+		return date.format( 'LL' );
+	}
+
+	getHref() {
+		const { period, path, siteSlug } = this.props;
+
+		// Some modules do not have view all abilities
+		if ( period && path && siteSlug ) {
+			return (
+				`/stats/${ period.period }/${ path }/${ siteSlug }` +
+				`?startDate=${ period.startOf.format( 'YYYY-MM-DD' ) }`
+			);
+		}
+	}
+}
+
+const yearsAgoQuery = ( startDate, endDate, yearsAgo = 1 ) => ( {
+	after: startDate.clone().subtract( yearsAgo, 'years' ).format(),
+	before: endDate.clone().subtract( yearsAgo, 'years' ).format(),
+	status: 'publish',
+} );
+
+export default connect( () => {
+	/*
+	 * Keep track of the posts queries and results.
+	 *
+	 * This is needed because <QueryPosts /> handles one query at a time and is
+	 * responsible of creating the data in the Redux store.
+	 *
+	 * Creating these arrays on the fly would create a new reference every time,
+	 * which would trigger a re-rendering of the component because we need to
+	 * pass them to it.
+	 *
+	 * previousPeriod is used to recreate these arrays when the period has changed.
+	 */
+	let postsByYear = [];
+	let postsByYearQueries = [];
+	let previousPeriod = null;
+
+	const oldestPostQuery = { number: 1, order: 'ASC' };
+
+	const mapStateToProps = ( state, props ) => {
+		const siteId = getSelectedSiteId( state );
+		const siteSlug = getSiteSlug( state, siteId );
+		const oldestPostResult = getSitePostsForQuery( state, siteId, oldestPostQuery );
+		const oldestPost = oldestPostResult && oldestPostResult[ 0 ];
+		const { summary, period } = props;
+
+		if ( period !== previousPeriod ) {
+			postsByYear = [];
+			postsByYearQueries = [];
+			previousPeriod = period;
+		}
+
+		/* After having received the oldest post date, an array of queries is
+		 * created: one for each year preceding the current year.
+		 *
+		 * Things to note here:
+		 *  - The third argument of the diff() method is set to true to get a
+		 *    floating point instead of a rounded number.
+		 *  - Math.floor() is used to get the amount of years as an integer.
+		 */
+		if ( ! postsByYearQueries.length && oldestPost && oldestPost.date ) {
+			const years = Math.floor(
+				Math.min(
+					MAX_POST_QUERIES,
+					period.startOf.diff( moment( oldestPost.date ).startOf( 'day' ), 'years', true ),
+				),
+			);
+			if ( years > 0 ) {
+				postsByYearQueries = [ ...Array( years + 1 ) ].map(
+					( v, i ) => yearsAgoQuery( period.startOf, period.endOf, i + 1 ),
+				);
+			}
+		}
+
+		// Fill postsByYear if any item has been updated
+		postsByYearQueries.forEach( ( query, i ) => {
+			const result = getSitePostsForQuery( state, siteId, query );
+
+			// Create a new array when an item is updated to invalidate
+			// the equality check and render the component.
+			if ( postsByYear[ i ] !== result ) {
+				postsByYear[ i ] = result || [];
+				postsByYear = [ ...postsByYear ];
+			}
```

Comment:
I find this code section confusing. Please consider reworking to make it a bit more understandable at a glance.